server/src/lib/migrate.rs

```diff
@@ -0,0 +1,308 @@
+    io::{AsyncReadExt, AsyncWriteExt},
+    task::JoinHandle,
+};
+use uuid::Uuid;
+
+use crate::server::Context;
+
+/// Our migration protocol version
+const MIGRATION_PROTOCOL_VERION: usize = 0;
+
+/// Our migration protocol encoding
+const MIGRATION_PROTOCOL_ENCODING: ProtocolEncoding = ProtocolEncoding::Ron;
+
+/// The concatenated migration protocol-encoding-version string
+const MIGRATION_PROTOCOL_STR: &'static str = const_format::concatcp!(
+    "propolis-migrate-",
+    encoding_str(MIGRATION_PROTOCOL_ENCODING),
+    "/",
+    MIGRATION_PROTOCOL_VERION
+);
+
+/// Supported encoding formats
+enum ProtocolEncoding {
+    Ron,
+}
+
+/// Small helper function to stringify ProtocolEncoding
+const fn encoding_str(e: ProtocolEncoding) -> &'static str {
+    match e {
+        ProtocolEncoding::Ron => "ron",
+    }
+}
+
+/// Handle and context for an ongoing migration from a source instance
+pub struct MigrateSourceTask {
+    _task: JoinHandle<()>,
+}
+
+/// Errors which may occur during the course of a migration
+#[derive(Error, Debug)]
+pub enum MigrateError {
+    #[error("{0}")]
+    Http(#[from] hyper::Error),
+
+    #[error("couldn't establish migration connection to source instance")]
+    Initiate,
+
+    #[error("the source ({0}) and destination ({1}) instances are incompatible for migration")]
+    Incompatible(String, String),
+
+    #[error("expected connection upgrade")]
+    UpgradeExpected,
+
+    #[error("destination instance already initialized")]
+    DestinationAlreadyInitialized,
+
+    #[error("source instance is not initialized")]
+    SourceNotInitialized,
+
+    #[error("unexpected Uuid")]
+    UuidMismatch,
+
+    #[error("a migration from the current instance is already in progress")]
+    MigrationAlreadyInProgress,
+
+    #[error("protocol error")]
+    // TODO: just for testing rn
+    Protocol,
+}
+
+impl MigrateError {
+    fn incompatible(src: &str, dst: &str) -> MigrateError {
+        MigrateError::Incompatible(src.to_string(), dst.to_string())
+    }
+}
+
+impl Into<HttpError> for MigrateError {
+    fn into(self) -> HttpError {
+        let msg = format!("migration failed: {}", self);
+        match &self {
+            MigrateError::Http(_)
+            | MigrateError::Initiate
+            | MigrateError::Incompatible(_, _)
+            | MigrateError::DestinationAlreadyInitialized
+            | MigrateError::SourceNotInitialized
+            | MigrateError::UuidMismatch => HttpError::for_internal_error(msg),
+            MigrateError::MigrationAlreadyInProgress
+            | MigrateError::Protocol
+            | MigrateError::UpgradeExpected => {
+                HttpError::for_bad_request(None, msg)
+            }
+        }
+    }
+}
+
+async fn source_migrate_task(
+    _instance: Arc<Instance>,
+    mut conn: Upgraded,
+    log: slog::Logger,
+) -> Result<(), MigrateError> {
+    info!(log, "Enter Migrate Task");
+
+    // TODO: actual migration protocol, for now just send some stuff back and forth
```

Comment:
cc @dancrossnyc this would be where we plug in the tokio codec Framed stuff to wrap this `Upgraded` object.

Reply:
Cool; thanks.